:::{.column-body-outset}
<h1 style="color: #00205b; font-family: 'Georgia';">Forecasting Demand: Unveiling the Future of Product Sales and Crafting the Optimal Inventory Strategy</h1>
<author style="color: #00205b; font-family: 'Georgia';">By: <a href="https://www./">Simon Tagbor</a></author>
<date style="color: #00205b; font-family: 'Georgia';">2024-01-21</date>
:::


<blockquote class="blockquote">
<p>
One of the core focus of downstream supply chain operations involves a key responsibility – the ability to navigate the ever-changing landscape of customer demand. Supply chain professionals have adopted statiscal techqniques that ebable them to infer the demand of products based on historical data. This process is relatively straightforward when dealing with smaller product categories and stable demand patterns. However, challenges arise in larger product categories with constantly shifting consumer demands, posing scalability issues for traditional forecasting methods.
</p>
<p>
The traditional approach involves observing the distribution and variability of historical data from ERP systems, combined with expert intuition. Yet, as businesses grow, reducing guesswork in demand forecasting becomes important. The decisions made in inventory optimization can significantly impact costs.
</p>
<p>
In this notebook, I will guide you through building a demand forecasting model to predict product demand based on historical data. Moreover, I’ll demonstrate how this model can be leveraged for inventory optimization, covering concepts like reorder points, safety stock, and economic order quantity (EOQ).
</p>
<p>
Throughout this project, we will address two key business questions:
</p>
<ol>
<li>
<em>What is the demand forecast for the top selling product in the next 24 months?</em>
</li>
<li>
<em>What is the optimal inventory level for the product?</em>
</li>
</ol>
<p>
Join me on this exploration as we seek answers, and feel free to share your thoughts and feedback on my approach.
</p>
<p>You can find the source code for this project at my <a href="https://github.com/Simontagbor/supply-chain-demand-forecast">github page.</a> You can also find the Jupyter notebook and the dataset on my <a href="https://www.kaggle.com/simontagbor/british-airways-predictive-ml">kaggle page</a>.</p>
</blockquote>

<details>
<summary style="font-weight:bold;">
Click to Read Executive Summary on Research Findings
</summary>
<div style="
    font-size: 1.8vw; 
    color: #333; 
    padding: 1.5vw; 
    margin: 1.7vw 0; 
    border-left: 0.35vw solid #7792f9; 
    line-height: 1.6; 
    position: relative; 
    background: #e9f5ff;">
<h2 class="anchored">
Executive Summary
</h2>
<p>
After building and testing the demand forecasting model, we discovered trends, seasonalities and holiday effects on the top-selling product based on the dataset provided we also found the optimal inventory policy for the top-selling product:
</p>
<ul>
<li>
Given the zero variance observed in the product price, The Demand for the <code>product Card ID 365</code> is expected to remain fairly stable within the next two years with cyclical dips in sales within the third quarter of each year(2015 and 2016). it might be worth looking into these top predictors of demand outcomes to control the dips and ultimately improve sales outcomes
</li>
</ul>
<div style="display: grid; grid-template-columns: auto auto; padding: 10px; background-color: white; width:85%; margin-left: 9%;">
<div style="font-size: 1.5vw; text-align: left;">
<ol type="1">
<li>Pending
</li></ol></div>
<div style="font-size: 1.5vw; text-align: left;">
<ol start="2" type="1">
<li>Pending
</li></ol></div>
<div style="font-size: 1.5vw; text-align: left;">
<ol start="3" type="1">
<li>Pending
</li></ol></div>
<div style="font-size: 1.5vw; text-align: left;">
<ol start="4" type="1">
<li>Pending
</li></ol></div>
<div style="font-size: 1.5vw; text-align: left;">
<ol start="5" type="1">
<li>Pending
</li></ol></div>
<div style="font-size: 1.5vw; text-align: left;">
<ol start="6" type="1">
<li>Pending
</li></ol></div>
<div style="font-size: 1.5vw; text-align: left;">
<ol start="7" type="1">
<li>Pending
</li></ol></div>
<div style="font-size: 1.5vw; text-align: left;">
<ol start="8" type="1">
<li>Pending
</li></ol></div>
<div style="font-size: 1.5vw; text-align: left;">
<ol start="9" type="1">
<li>Pending
</li></ol></div>
<div style="font-size: 1.5vw; text-align: left;">
<ol start="10" type="1">
<li>Pending
</li></ol></div>
</div>
<ul>
<li>
Based on the Economic Order Quantity Model(EOQ) the optimal inventory policy is
</li>
<p>
</p>
</ul>
<h2 class="anchored">
Potential Application of Insights
</h2>
<p>
</p>
<p>
</p>
</div>
</details>

<h2>Project Outline</h2>
<p>For this project, I completed the following tasks:</p>

- [x] Performed Exploratory Data Analysis.
- [x] Cleaned and Prepared the data for modeling.
- [x] Conducted Time Series Modeling With Prophet.
- [x] Evaluated the model performance.
- [x] Interpret the model results and answer the business questions.


<h2>Problem Statement</h2>
<em>Large product categories and constantly shifting consumer demand patterns introduce a scaling challenge for traditional demand forecasting techniques. There is a need for an approach that reduces the level of guesswork and reduces the avoidable costly outcomes of poor inventory optimizations.</em>

<h2>Project Dependencies</h2>


In [1]:
#| code fold true
# import project libraries
import pandas as pd
import numpy as np # for linear algebra
import math # for math operations 

import seaborn as sns # for plotting

# handling files
import os 
import sys 

# data preprocessing
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split


# Model Building and Fitting
from sklearn.ensemble import RandomForestClassifier
from prophet import Prophet



# Model Evaluation and Tuning
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# visualisation libraries
import seaborn as sns
import matplotlib.pyplot as plt
import matplotlib.pyplot as plt # for plotting
import squarify # for tree maps

<h1>Eploratory Data Analysis</h1>
<p>
The working <a href="#">dataset</a> contains entries of customer demand information The data contains 53 features (columns)
</p>
<p>
To understand the data, I performed exploratory data analysis. I used the following techniques to understand the data:
</p>
<ul class="task-list">
<li><input type="checkbox" checked="">Visual inspection of data.</li>
<li><input type="checkbox" checked="">Exploratory Data Visualizations. (Univariate and Bivariate)</li>
</ul>

<h2>Visual Inspection of Data</h2>

In [2]:
df = pd.read_csv("data/DataCoSupplyChainDataset.csv", encoding="ISO-8859-1")
df.head(2)

Type  Days for shipping (real)  Days for shipment (scheduled)  \
0     DEBIT                         3                              4   
1  TRANSFER                         5                              4   

   Benefit per order  Sales per customer   Delivery Status  \
0          91.250000          314.640015  Advance shipping   
1        -249.089996          311.359985     Late delivery   

   Late_delivery_risk  Category Id   Category Name Customer City  ...  \
0                   0           73  Sporting Goods        Caguas  ...   
1                   1           73  Sporting Goods        Caguas  ...   

  Order Zipcode Product Card Id Product Category Id  Product Description  \
0           NaN            1360                  73                  NaN   
1           NaN            1360                  73                  NaN   

                                  Product Image  Product Name Product Price  \
0  http://images.acmesports.sports/Smart+watch   Smart watch         327.75   
1  http://images.acmesports.sports/Smart+watch   Smart watch         327.75   

  Product Status shipping date (DateOrders)   Shipping Mode  
0              0             2/3/2018 22:56  Standard Class  
1              0            1/18/2018 12:27  Standard Class  

[2 rows x 53 columns]

<p>To explore the spread of the data, we will use the describe() method to get the summary statistics of the data.</p>

In [14]:
# retrieve the number of columns and rows
df.describe()

Days for shipping (real)  Sales per customer  Late_delivery_risk  \
count             180519.000000       180519.000000       180519.000000   
mean                   3.497654          183.107609            0.548291   
min                    0.000000            7.490000            0.000000   
25%                    2.000000          104.379997            0.000000   
50%                    3.000000          163.990005            1.000000   
75%                    5.000000          247.399994            1.000000   
max                    6.000000         1939.989990            1.000000   
std                    1.623722          120.043670            0.497664   

               Sales  Product Card Id  Product Category Id  Product Price  \
count  180519.000000    180519.000000        180519.000000  180519.000000   
mean      203.772096       692.509764            31.851451     141.232550   
min         9.990000        19.000000             2.000000       9.990000   
25%       119.980003       403.000000            18.000000      50.000000   
50%       199.919998       627.000000            29.000000      59.990002   
75%       299.950012      1004.000000            45.000000     199.990005   
max      1999.989990      1363.000000            76.000000    1999.989990   
std       132.273077       336.446807            15.640064     139.732492   

       Product Status     shipping date (DateOrders)  Shipping Year  \
count        180519.0                         180519  180519.000000   
mean              0.0  2016-06-16 05:45:23.202433024    2015.979404   
min               0.0            2015-01-03 00:00:00    2015.000000   
25%               0.0            2015-09-25 06:59:00    2015.000000   
50%               0.0            2016-06-15 08:32:00    2016.000000   
75%               0.0            2017-03-04 21:29:00    2017.000000   
max               0.0            2018-02-06 22:14:00    2018.000000   
std               0.0                            NaN       0.831962   

       Shipping Month   Shipping Day  Shipping Weekday  
count   180519.000000  180519.000000     180519.000000  
mean         6.252494      15.702170          2.997385  
min          1.000000       1.000000          0.000000  
25%          3.000000       8.000000          1.000000  
50%          6.000000      16.000000          3.000000  
75%          9.000000      23.000000          5.000000  
max         12.000000      31.000000          6.000000  
std          3.398391       8.808808          2.002633

<details><summary>Click to see some Notable Observation of the Data</summary>
<ol type="1">
<li><p>Aproximately <code>55%</code> of orders had late delivery risks.</p></li>
<li><p>Aproximately <code>75%</code> of products cost <code>$199.99</code></p></li>
<li><p>All the products are available.</p></li>
<li><p>75% of customers bought goods worth at least <code>$247.40</code></p></li>
</ol>
<p>
Further inspection of the data will help us understand the data better.
</p>
</details>


<h2 class="anchored">
Data Preprocessing
</h2>
<p>
To Forecast the demand based on the available data, we will focus on historical sales data, and product attributes like; stock level, and product category, we will also analyze the impact of other variables that contribute to demand patterns including geographic factors, customer segments and lead time.
</p>
<h4 class="anchored">
Preprocessing Tasks
</h4>

- [x] Drop irrelevant columns
- [x] Drop rows with missing values
- [x] Create new features
- [x] Convert categorical features to numerical features

<p>Based on the above, we will drop the majority of the columns that are not relevant for forecasting the demand and extract new features from the existing columns
</p>



<h3>
Drop Irrelevant Columns
</h3>

In [4]:
#| code fold true
# drop irrelevant columns
def drop_columns(df, columns_to_drop):
    try:
        df = df.drop(columns=columns_to_drop)
        print(f"{len(columns_to_drop)} columns dropped successfully. Number of columns remaining: {len(df.columns)}")
        return df
    except KeyError as e:
        print(f"""Column(s): {e} not found in dataframe.
              
            No columns dropped.
            Please Check that the column names are correct.""")
        return df

# Specify the columns to keep
colums_to_keep = ['Days for shipping (real)', 
                  'Customer Country',
                  'Sales per customer',
                  'Delivery Status', 
                  'Late_delivery_risk', 
                  'Customer City',
                  'Customer Segment',
                  'Sales','Shipping Mode',
                  'Type', 'Product Card Id',
                  'Customer Zipcode', 
                  'Product Category Id', 
                  'Product Name',                    
                  'Product Price',
                  'Market', 
                  'Product Status',
                  'shipping date (DateOrders)',]

# Specify the columns to drop
columns_to_drop = [col for col in df.columns if col not in colums_to_keep ]

df = drop_columns(df, columns_to_drop)

35 columns dropped successfully. Number of columns remaining: 18


<h3>Drop Rows with Missing Values</h3>

In [7]:
#| code fold true
# drop customer Zip code.
df = df.drop(columns=['Customer Zipcode'])

<h4>A Quick Spot Check for Missing Values</h4>

In [8]:
### Check for Missing values
df.isnull().sum()

Type                          0
Days for shipping (real)      0
Sales per customer            0
Delivery Status               0
Late_delivery_risk            0
Customer City                 0
Customer Country              0
Customer Segment              0
Market                        0
Sales                         0
Product Card Id               0
Product Category Id           0
Product Name                  0
Product Price                 0
Product Status                0
shipping date (DateOrders)    0
Shipping Mode                 0
Shipping Year                 0
Shipping Month                0
Shipping Day                  0
Shipping Weekday              0
dtype: int64

<h3>Create New Features</h3>
<p>The dataset contains a `shipping date` column which is a `DateTime` object from which we can extract `Month`, `Year`, `Day` and `Day of Week` that can be useful in our analysis.</p>

- [x] `Month` - to capture the months per sale.
- [x] `Year` - to capture the year per sales.
- [x] `Day` - to capture the day per sales.
- [x] `Day of Week` - to capture the day of the week per sales.

<p>
I chose <code>Shipping Date</code> over the <code>Order date</code> because it reflects fulfilled demand while the latter captures interest. (the customer could cancel the order)
</p>

In [5]:
#| code fold true
# Create month, Year, Day, and Weekday columns from Shipping Date
def extract_date_parts(df, date_column, prefix):
    try:
        df[date_column] = pd.to_datetime(df[date_column])
        df[f'{prefix} Year'] = df[date_column].dt.year
        df[f'{prefix} Month'] = df[date_column].dt.month
        df[f'{prefix} Day'] = df[date_column].dt.day
        df[f'{prefix} Weekday'] = df[date_column].dt.weekday
        # verify and notify that the columns have been created
        if f'{prefix} Year' in df.columns and f'{prefix} Month' in df.columns and f'{prefix} Day' in df.columns and f'{prefix} Weekday' in df.columns:
            print(f" Success! Columns Created: {prefix} Year, {prefix} Month, {prefix} Day, and {prefix} Weekday")
            return df
        else:
            print("Error creating columns. Please check that the date column name is correct.")
    except Exception as e:
        print(f"Error creating columns: {e}")
        return df

# Use the function to extract date parts
df = extract_date_parts(df, 'shipping date (DateOrders)', 'Shipping')

 Success! Columns Created: Shipping Year, Shipping Month, Shipping Day, and Shipping Weekday


In [6]:
# display the shape of the dataframe
df.shape

(180519, 22)

<p>Now we have 22 new columns and 180519 entries (rows) in the dataset.</p>

<h3>Convert Categorical Features to Numerical Features</h3>
<p>The nature of categorical data makes it unsuitable for future analysis. for instance, machine learning models can't work with categorical values for customer origins like `UK`, `USA`, `France`, etc. We will convert these categorical values to numerical values using the `LabelEncoder` from the `sklearn` library.</p>

<p>I will also perform a <a href="#"> one-hot encoding</a> technique on categorical features for future machine learning modeling tasks.</p>

<p> I wrote a `prepare_data()` function that returns two preprocessed dataframes: one that is encoded using a label encoder function and the other encoded using one hot encoding technique.</p>

<p>You can learn about encoding techniques for categorical variables <a href="#">here</a></p>


In [11]:
# Select top selling product
top_product = df['Product Card Id'].value_counts().index[0]
# get top product ID
print(f"Filtering and Encoding Dataset for Top Product ID: {top_product}")

from sklearn.preprocessing import LabelEncoder

def prepare_data(df, product_card_id, categorical_cols, columns_to_drop):
    """
    Prepare a DataFrame for bivariate analysis and machine learning by applying label encoding and one-hot encoding to categorical columns and dropping specified columns.

    Parameters:
    df (pandas.DataFrame): The original DataFrame.
    product_card_id (int): The product card ID to filter the DataFrame on.
    categorical_cols (list of str): The names of the categorical columns to apply encoding to.
    columns_to_drop (list of str): The names of the columns to drop from the DataFrame.

    Returns:
    pandas.DataFrame: The label encoded DataFrame for bivariate analysis.
    pandas.DataFrame: The one-hot encoded DataFrame for machine learning.
    """
    try:
        df_copy = df[df['Product Card Id'] == product_card_id].copy()  # create a copy

        # label encoding
        label_encoder = LabelEncoder()
        df_label_encoded = df_copy.copy()

        # Apply label encoding to categorical variables in place
        for col in categorical_cols:
            df_label_encoded[col] = label_encoder.fit_transform(df_label_encoded[col])

        # Drop specified columns
        df_label_encoded = df_label_encoded.drop(columns=columns_to_drop)

        # one-hot encoding
        df_one_hot_encoded = pd.get_dummies(df_copy, columns=categorical_cols)

        # Drop specified columns
        df_one_hot_encoded = df_one_hot_encoded.drop(columns=columns_to_drop)
        print("Data preparation successful.")
        return  df_one_hot_encoded, df_label_encoded
    except Exception as e:
        print(f"Error preparing data: {e}")
        return None, None

# Use the function to prepare the data for bivariate analysis
categorical_cols = ['Type', 'Customer Segment', 
                    'Delivery Status', 
                    'Customer City', 
                    'Market',
                    'Shipping Mode']

columns_to_drop = ['Product Name', 
                   'Customer Country', 
                   'shipping date (DateOrders)', 
                   'Product Card Id', 
                   'Product Category Id', 
                   'Product Status', 
                   'Product Price']

# drop columns and encode data for correlation martrix and Machine learning
onehot_encode_df, label_encode_df = prepare_data(df, top_product, categorical_cols, columns_to_drop)

# rename Type column to Payment Type
label_encode_df = label_encode_df.rename(columns={'Type': 'Payment Type'})
onehot_encode_df = onehot_encode_df.rename(columns={'Type': 'Payment Type'})

Filtering and Encoding Dataset for Top Product ID: 365
Data preparation successful.


<h4>Confirm Encoding of Dataset </h4>

In [12]:
# validate the label encoding
label_encode_df.dtypes

Payment Type                  int64
Days for shipping (real)      int64
Sales per customer          float64
Delivery Status               int64
Late_delivery_risk            int64
Customer City                 int64
Customer Segment              int64
Market                        int64
Sales                       float64
Shipping Mode                 int64
Shipping Year                 int32
Shipping Month                int32
Shipping Day                  int32
Shipping Weekday              int32
dtype: object

In [13]:
# validate the one-hot encoding
onehot_encode_df.dtypes

Days for shipping (real)          int64
Sales per customer              float64
Late_delivery_risk                int64
Sales                           float64
Shipping Year                     int32
                                 ...   
Market_USCA                        bool
Shipping Mode_First Class          bool
Shipping Mode_Same Day             bool
Shipping Mode_Second Class         bool
Shipping Mode_Standard Class       bool
Length: 590, dtype: object

<h4>Data Preprocessing Completed Sucessfully!!</h4>


<!-- insert gif -->
<img src="https://media.giphy.com/media/jIv6pfqKiIvHPYZO6y/giphy.gif?cid=790b76114sdadclwy3mcxlm3o08nf63hxm28relnkbokg8ue&ep=v1_gifs_search&rid=giphy.gif&ct=g" alt="gif" style="width: 100%; height: auto;">	

<p> The dataset is now ready for further analysis and modeling. I will now proceed to conduct exploratory data visualizations to understand the data better.</p>

<h2>Exploratory Data Visualizations</h2>
<p>To highlight the distributions of the individual variables as well as the relationship between the variables and the target variables, I used the following techniques:</p>

- [x] Univariate Analysis
- [x] Exploratory Time Series Analysis

<h3>Univariate Analysis</h3>
<p>Univariate analysis is the simplest form of data analysis where the data being analyzed contains only one variable. Since it's a single variable, it doesn't deal with causes or relationships. The main purpose of univariate analysis is to describe the data and find patterns that exist within it.</p>

<h4>Visualizing the Distribution of the Dataset</h4>